testing/proxytest/https.go

```diff
@@ -0,0 +1,220 @@
+			p.http500Error(clientTLSConn, "failed writing TLS response header", err, log)
+			return
+		}
+
+		// 3rd - indicates the headers are done and the body will follow
+		if _, err = clientTLSConn.Write([]byte("\r\n")); err != nil {
+			p.http500Error(clientTLSConn, "failed writing TLS header/body separator", err, log)
+			return
+		}
+
+		// copy the body else
+		_, err = io.CopyBuffer(clientTLSConn, resp.Body, make([]byte, 4096))
+		if err != nil {
+			p.http500Error(clientTLSConn, "failed writing response body", err, log)
+			return
+		}
+
+		_ = resp.Body.Close()
+	}
+
+	log.Debug("EOF reached, finishing HTTPS handler")
+}
+
+func (p *Proxy) newTLSCert(u *url.URL) (*tls.Certificate, error) {
+	// generate the certificate key - it needs to be RSA because Elastic Defend
+	// do not support EC :/
+	priv, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, fmt.Errorf("could not create RSA private key: %w", err)
+	}
+	host := u.Hostname()
+
+	var name string
+	var ips []net.IP
+	ip := net.ParseIP(host)
+	if ip == nil { // host isn't an IP, therefore it must be an DNS
+		name = host
+	} else {
+		ips = append(ips, ip)
+	}
+
+	cert, _, err := certutil.GenerateGenericChildCert(
+		name,
+		ips,
+		priv,
+		&priv.PublicKey,
+		p.ca.capriv,
+		p.ca.cacert)
+	if err != nil {
+		return nil, fmt.Errorf("could not generate TLS certificate for %s: %w",
+			host, err)
+	}
+
+	return cert, nil
+}
+
+func (p *Proxy) http500Error(clientCon net.Conn, msg string, err error, log *slog.Logger) {
+	p.httpError(clientCon, http.StatusInternalServerError, msg, err, log)
+}
+
+func (p *Proxy) httpError(clientCon net.Conn, status int, msg string, err error, log *slog.Logger) {
+	log.Error(msg, "err", err)
+
+	_, err = clientCon.Write(generateHTTPResponse(status, []byte(msg)))
+	if err != nil {
+		log.Error("failed writing response", "err", err)
+	}
+}
+
+func hijack(w http.ResponseWriter) (net.Conn, error) {
+	hijacker, ok := w.(http.Hijacker)
+	if !ok {
+		w.WriteHeader(http.StatusInternalServerError)
+		_, _ = fmt.Fprint(w, "cannot handle request")
+		return nil, errors.New("http.ResponseWriter does not support hijacking")
+	}
+
+	clientCon, _, err := hijacker.Hijack()
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		_, err = fmt.Fprint(w, "cannot handle request")
+
+		return nil, fmt.Errorf("could not Hijack HTTPS CONNECT request: %w", err)
+	}
+
+	return clientCon, err
+}
+
+func cleanUpHeaders(h http.Header) {
+	h.Del("Proxy-Connection")
+	h.Del("Proxy-Authenticate")
+	h.Del("Proxy-Authorization")
+	h.Del("Connection")
+}
+
+func generateHTTPResponse(statusCode int, body []byte) []byte {
+	resp := bytes.Buffer{}
+	resp.WriteString(fmt.Sprintf("HTTP/1.1 %d %s\r\n",
+		statusCode, http.StatusText(statusCode)))
+	resp.WriteString("Content-Type: text/plain\r\n")
+	resp.WriteString(fmt.Sprintf("Content-Length: %d\r\n", len(body)))
+	resp.WriteString("\r\n")
+	if len(body) > 0 {
+		resp.Write(body)
+	}
```

Comment:
This would look better if we instead created a `http.Response` and written it to the buffer.